geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/index/IndexKeySpace.scala

```diff
@@ -59,4 +63,99 @@ trait IndexKeySpace[T] {
     * @return
     */
   def getIndexValues(sft: SimpleFeatureType, filter: Filter, explain: Explainer): T
+
+  /**
+    * Creates ranges over the index keys
+    *
+    * @param values index values @see getIndexValues
+    * @return
+    */
+  def getRanges(values: T): Iterator[ScanRange[U]]
+
+  /**
+    * Creates bytes from ranges
+    *
+    * @param ranges typed scan ranges. @see `getRanges`
+    * @param prefixes prefixes to the range, if any. each prefix will create a new range
+    * @return
+    */
+  def getRangeBytes(ranges: Iterator[ScanRange[U]],
+                    prefixes: Seq[Array[Byte]] = Seq.empty,
+                    tier: Boolean = false): Iterator[ByteRange]
+
+  /**
+    * Determines if the ranges generated by `getRanges` are sufficient to fulfill the query,
+    * or if additional filtering needs to be done
+    *
+    * @param config data store config
+    * @param values index values @see getIndexValues
+    * @param hints query hints
+    * @return
+    */
+  def useFullFilter(values: Option[T],
+                    config: Option[GeoMesaDataStoreConfig],
+                    hints: Hints): Boolean
+}
+
+object IndexKeySpace {
+
+  type ToIndexKeyBytes = (Seq[Array[Byte]], SimpleFeature, Array[Byte]) => Seq[Array[Byte]]
+
+  sealed trait ByteRange
+
+  case class BoundedByteRange(lower: Array[Byte], upper: Array[Byte]) extends ByteRange
+  case class SingleRowByteRange(row: Array[Byte]) extends ByteRange
+  case class TieredByteRange(lower: Array[Byte],
+                             upper: Array[Byte],
+                             lowerTierable: Boolean = false,
+                             upperTierable: Boolean = false) extends ByteRange
+
+  object ByteRange {
+
+    import org.locationtech.geomesa.utils.conversions.ScalaImplicits.RichTraversableOnce
+
+    implicit val ByteOrdering: Ordering[Array[Byte]] =
+      Ordering.comparatorToOrdering(UnsignedBytes.lexicographicalComparator)
+
+    val UnboundedLowerRange: Array[Byte] = Array.empty
+    val UnboundedUpperRange: Array[Byte] = Array.fill(3)(ByteArrays.MaxByte)
+
+    def ranges(prefixes: Seq[Array[Byte]], row: Array[Byte]): Iterator[ByteRange] = {
```

Comment:
Just to clarify... the prefixes here are actually *distinct* prefixes correct?  (I.e., not a sequence of byte array to combine into one prefix?)

Reply:
right - i'll add more scaladocs :)

Reply:
actually it doesn't seem that these are actually used - i'm going to remove them